File: src/main/kotlin/ninja/blacknet/crypto/Message.kt
/*
 * Copyright (c) 2018-2020 Pavel Vasin
 *
 * Licensed under the Jelurida Public License version 1.1
 * for the Blacknet Public Blockchain Platform (the "License");
 * you may not use this file except in compliance with the License.
 * See the LICENSE.txt file at the top-level directory of this distribution.
 */

package ninja.blacknet.crypto

import ninja.blacknet.MESSAGE_SIGN_NAME
import ninja.blacknet.crypto.Blake2b.buildHash

object Message {
    private const val SIGN_MAGIC = "$MESSAGE_SIGN_NAME Signed Message:\n"

    fun sign(privateKey: ByteArray, message: String): ByteArray {
        return Ed25519.sign(hash(SIGN_MAGIC, message), privateKey)
    }

    fun verify(publicKey: ByteArray, signature: ByteArray, message: String): Boolean {
        return verifyWithMagic(publicKey, signature, message, SIGN_MAGIC)
    }

    internal fun verifyWithMagic(publicKey: ByteArray, signature: ByteArray, message: String, magic: String): Boolean {
        return Ed25519.verify(signature, hash(magic, message), publicKey)
    }

    private fun hash(magic: String, message: String): ByteArray {
        return buildHash {
            encodeString(magic)
            encodeString(message)
        }
    }
}

File: network/src/main/rust/wallet/message.rs
/*
 * Copyright (c) 2018-2026 Pavel Vasin
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

use blacknet_compat::Mode;
use blacknet_crypto::symmetric::Blake2b256;
use blacknet_kernel::{
    blake2b::Hash256,
    ed25519::{PublicKey, SecretKey, Signature, sign, verify},
    error::Result,
};

const SIGN_MAGIC: &str = " Signed Message:\n";

pub fn sign_message(mode: &Mode, secret_key: &SecretKey, message: &str) -> Signature {
    sign(hash(mode, message), secret_key)
}

pub fn verify_message(
    mode: &Mode,
    public_key: PublicKey,
    signature: Signature,
    message: &str,
) -> Result<()> {
    verify(signature, hash(mode, message), public_key)
}

fn hash(mode: &Mode, message: &str) -> Hash256 {
    let mut hasher = Blake2b256::new();
    hasher.update(mode.message_sign_name());
    hasher.update(SIGN_MAGIC);
    hasher.update(message);
    hasher.finalize().into()
}
